File: uplift-gradle-plugins/src/main/java/com/github/kjetilv/uplift/plugins/UpliftCdkTask.java
package com.github.kjetilv.uplift.plugins;

import com.github.kjetilv.uplift.plugins.core.CdkApp;
import org.gradle.api.artifacts.Configuration;
import org.gradle.api.artifacts.Dependency;
import org.gradle.api.provider.Property;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.InputFile;
import org.gradle.api.tasks.PathSensitive;
import org.gradle.api.tasks.PathSensitivity;
import org.gradle.work.DisableCachingByDefault;

import java.nio.file.Path;
import java.util.List;
import java.util.Set;

/** Base for the tasks that generate the throwaway CDK application. */
@DisableCachingByDefault(because = "Deploying a stack is not cacheable")
public abstract class UpliftCdkTask extends UpliftTask {

    /** CDK brings its own copies of these, so they are not injected into the generated pom. */
    private static final Set<String> AWS = Set.of(
        "software.amazon.awscdk:aws-cdk-lib",
        "software.constructs:constructs"
    );

    @InputFile
    @PathSensitive(PathSensitivity.RELATIVE)
    public abstract Property<Path> getStackbuilderJar();

    @Input
    public abstract Property<String> getStackbuilderClass();

    @Override
    public void stackWith(String name) {
        getStackbuilderClass().set(name);
    }

    protected final void initCdkApp() {
        new CdkApp(
            cdk(),
            getStackbuilderJar().getOrNull(),
            getStackbuilderClass().getOrElse(""),
            getAccount().get(),
            getRegion().get(),
            getStack().get(),
            dependencies()
        ).initialize();
    }

    private List<CdkApp.Dependency> dependencies() {
        Configuration compileClasspath = getProject().getConfigurations().findByName("compileClasspath");
        if (compileClasspath == null) {
            return List.of();
        }
        return compileClasspath.getIncoming().getDependencies().stream()
            .filter(dependency -> !isAws(dependency))
            .map(dependency -> new CdkApp.Dependency(
                dependency.getGroup(),
                dependency.getName(),
                dependency.getVersion()))
            .toList();
    }

    private static boolean isAws(Dependency dependency) {
        return AWS.contains(dependency.getGroup() + ":" + dependency.getName());
    }
}

File: uplift-gradle-plugins/src/main/kotlin/com/github/kjetilv/uplift/plugins/UpliftCdkTask.kt
package com.github.kjetilv.uplift.plugins

import org.gradle.api.artifacts.Dependency
import org.gradle.api.provider.Property
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.InputFile
import org.gradle.api.tasks.PathSensitive
import org.gradle.api.tasks.PathSensitivity
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.util.stream.Collectors

abstract class UpliftCdkTask : UpliftTask() {

    @get:InputFile
    @get:PathSensitive(PathSensitivity.RELATIVE)
    abstract val stackbuilderJar: Property<Path>

    @get:Input
    abstract val stackbuilderClass: Property<String>

    override fun stackWith(name: String) {
        this.apply {
            stackbuilderClass %= name
        }
    }

    protected fun initCdkApp() {
        clearCdkApp()
        runCdk("cdk init --language=java --generate-only")
        copyTo(resolvedStackBuilderJar(), cdkApp())

        val writeCdkCode = loadResource("CloudApp.java").split('\n')
        val sourcePackage = cdkApp().resolve("src/main/java/lambda/uplift/app")
        Files.createDirectories(sourcePackage)
        Files.write(sourcePackage.resolve("CloudApp.java"), writeCdkCode)
        clearRecursive(listOf(cdkApp().resolve("src/main/java/com"), cdkApp().resolve("src/test/java/com")))

        val pom = cdkApp().resolve("pom.xml")
        val pomCopy = Files.copy(pom, cdkApp().resolve("pom.xml.orig"))
        Files.write(pom, templated(pomCopy))
        clearRecursive(pomCopy)
    }


    private fun templated(pomCopy: Path?) =
        Files.lines(pomCopy, StandardCharsets.UTF_8).collect(Collectors.toList()).flatMap { line ->
            if (line.contains("<mainClass>"))
                addedMain(indent(line), line)
            else if (line.contains("</dependencies>"))
                addedDeps(indent(line), line)
            else
                listOf(line)
        }

    private fun indent(main: String) = main.takeWhile(Char::isWhitespace)

    private fun addedMain(indent: String, line: String) = listOf(
        line.replace("com.myorg.AppApp", "lambda.uplift.app.CloudApp"),
        "$indent<systemProperties>",
        property("$indent  ", "uplift.account" to account.get()),
        property("$indent  ", "uplift.region" to region.get()),
        property("$indent  ", "uplift.stack" to stack.get()),
        property("$indent  ", "uplift.stackbuilderJar" to resolvedStackBuilderJar().fileName),
        property("$indent  ", "uplift.stackbuilderClass" to stackbuilderClass.get()),
        "$indent</systemProperties>"
    )

    private fun addedDeps(indent: String, line: String) =
        dependencies().filterNotNull().filterNot(::isAws).flatMap { dep ->
            dep.run {
                listOf(
                    "",
                    "$indent    <dependency>",
                    "$indent      <groupId>$group</groupId>",
                    "$indent      <artifactId>$name</artifactId>",
                    "$indent      <version>$version</version>",
                    "$indent    </dependency>"
                )
            }
        } + listOf(line)

    private fun dependencies() =
        project.configurations.findByName("compileClasspath")?.incoming?.dependencies?.stream()?.toList()?.toList()
            ?: emptyList()

    private fun property(indent: String, entry: Pair<String, Any>) =
        entry.let { (key, value) ->
            "$indent  <systemProperty><key>${key}</key><value>${value}</value></systemProperty>"
        }

    private fun resolvedStackBuilderJar() =
        stackbuilderJar.orNull ?: throw IllegalStateException("Required a stackbuilderJar")

    private fun isAws(dep: Dependency) =
        setOf(
            "software.amazon.awscdk" to "aws-cdk-lib",
            "software.constructs" to "constructs"
        ).contains(dep.group to dep.name)
}
